### PR TITLE
Sidekiq

### PR DESCRIPTION
**Pull request by fedelcar**
# Summary

Added worker form sidekiq that sends email in background
# Trello card

https://trello.com/c/w31WfYWX

### CODE REVIEW

**Comment by fedelcar** [comment]
@epintos Fixed comments from previous pull requests

File: .travis.yml [hunk 1]
```diff
@@ -16,10 +16,9 @@ services:
   - postgresql
 
 before_script:
-  - cp config/database.travis.yml config/database.yml
-  - psql -c "CREATE USER rails-bootstrap WITH PASSWORD 'rails-bootstrap';" -U postgres
-  - psql -c "DROP DATABASE IF EXISTS rails-bootstrap_test;" -U postgres
-  - psql -c "CREATE DATABASE rails-bootstrap_test ENCODING 'utf8';" -U postgres
+  - psql -c "CREATE USER federicolopezcanas WITH PASSWORD 'password';" -U postgres
```
Comment (epintos) [inline .travis.yml:19]: Its not bad that you used your user role. But if its a project for a client or one where there are many developers you can use the database.travis.yml for generic setup

File: Gemfile [hunk 2]
```diff
@@ -28,6 +28,9 @@ gem 'omniauth-google-oauth2'
 
 # Using ActiveAdmin
 gem 'activeadmin', github: 'activeadmin'
+# Use sidekiq for multithreading
+gem 'sidekiq'
+gem 'redis-rails'
```
Comment (epintos) [inline Gemfile:33]: I've never used this redis-rails gem. Not sure what extra support gives. Having redis installed in your ubuntu should be enought. Why did you add it? Did'nt work without it?

File: app/controllers/articles_controller.rb [hunk 3]
```diff
@@ -13,17 +13,25 @@ def new
 
   def send_last_10
     @articles = Article.last(10)
-    UserMailer.last_10_email(current_user,@articles).deliver
+    UserMailer.last_10_email(current_user.email, @articles).deliver_now
```
Comment (epintos) [inline app/controllers/articles_controller.rb:16]: `deliver_now` doesnt use sidekiq, right?

File: app/controllers/articles_controller.rb [hunk 4]
```diff
@@ -33,17 +41,15 @@ def create
 
   def update
     article
-
     if @article.update(article_params)
-      redirect_to @article
+      redirect_to @article if article.user = current_user.email
```
Comment (epintos) [inline app/controllers/articles_controller.rb:45]: I think it should be `article.user == current_user`

File: app/workers/hard_worker.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class HardWorker
+  include Sidekiq::Worker
+
+  def perform(user)
```
Comment (epintos) [inline app/workers/hard_worker.rb:4]: Be carefull with the methods and params names. Someone can confuse your param user with a user instance and in this case its an email.

File: app/workers/hard_worker.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    @articles = []
+    User.last(10).each do |u|
```
Comment (epintos) [inline app/workers/hard_worker.rb:6]: In Rails you have to be carefull with the each and the arrays. In this case you are iterating though little quantity of collections. But if there were many, you would be pushing lots of models into an array. This would alloc lots of memory. Another option would be doing all this in one huge sql query. But for now this works fine

File: app/controllers/articles_controller.rb (hunk 3 above)
Comment (epintos) [inline app/controllers/articles_controller.rb:16]: Now that you are using sidekiq. It would be nice to send all the emails through it. Be carefull that if you use sidekiq you can't send @articles. You should send the ids

File: app/controllers/articles_controller.rb [hunk 7]
```diff
@@ -13,17 +13,25 @@ def new
...
     @articles = Article.last(10)
-    UserMailer.last_10_email(current_user,@articles).deliver
+    UserMailer.last_10_email(current_user.email, @articles).deliver_now
     redirect_to root_path
   end
 
+  def send_last_10_by_last_5
+    # def get_content_to_display
+    HardWorker.perform_async(current_user.email)
+    render :processing_email do |page|
```
Comment (epintos) [inline app/controllers/articles_controller.rb:23]: I've never used this sintaxis. But I assume you are returning an html and loading it by ajax in the `remote_function`.

There is I think a better way to do this. The button could be a static button. And by jquery and ajax when you click it, you can make a request to `send_last_10_by_last_5`. This returns an html and by ajax you append it to your page.

**Comment by epintos** [comment]
Travis :x:

Comments
